syntax = "proto3";

package posts;

service PostService {
    rpc CreatePost (CreatePostRequest) returns (PostResponse) {}
    rpc DeletePost (DeletePostRequest) returns (PostResponse) {}
    rpc UpdatePost (UpdatePostRequest) returns (PostResponse) {}
    rpc GetPost (GetPostRequest) returns (PostResponse) {}
    rpc ListPosts (ListPostsRequest) returns (ListPostsResponse) {}
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    string creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message DeletePostRequest {
    int32 id = 1;
    string creator_id = 2;
}

message UpdatePostRequest {
    int32 id = 1;
    string title = 2;
    string description = 3;
    bool is_private = 4;
    repeated string tags = 5;
    string creator_id = 6;
}

message GetPostRequest {
    int32 id = 1;
    string creator_id = 2;
}

message ListPostsRequest {
    int32 page_number = 1;
    int32 page_size = 2;
    string creator_id = 3;
    string author_id = 4;
}

message PostResponse {
    bool success = 1;
    string message = 2;
    Post post = 3;
}

message ListPostsResponse {
    bool success = 1;
    string message = 2;
    repeated Post posts = 3;
}

message Post {
    int32 id = 1;
    string title = 2;
    string description = 3;
    string creator_id = 4;
    string created_at = 5;
    string updated_at = 6;
    bool is_private = 7;
    repeated string tags = 8;
}